### PR TITLE
Add EP with dropless

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

* Add functionality of EP with megablox/ragged_dot
* Add unit tests to ensure EP matches expected results
* Create a bug to track the performance improvement, b/403069924

# Tests

* unit test
* end-to-end test: 1) without the change [prof](https://xprof.corp.google.com/trace_viewer/ranran-13054269920223327014?hosts=t1v-n-1ddc60b1-w-0&host_index=0&trace_filter_config={}&view_start=606.464&view_end=7268.378); 2) with the change (slower performance) [prof](https://xprof.corp.google.com/trace_viewer/ranran-16868456941204155188?hosts=t1v-n-1ddc60b1-w-0&host_index=0&trace_filter_config={}&view_start=7629.044&view_end=7731.721) - see details about perf in b/403069924

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/linears.py [hunk 1]
```diff
@@ -422,6 +430,39 @@ def unpermute(self, intermediate, sorted_selected_experts, weights, batch_size,
   def sparse_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
     tile_size = (512, 1024, 1024)  # (m, k, n)
 
+    def transform_array(input_array, shard_id, strategy):
```
Comment (gobbleturk) [inline MaxText/layers/linears.py:433]: Can you include a comment describing what this function does?

File: MaxText/layers/linears.py [hunk 2]
```diff
@@ -463,14 +504,22 @@ def gmm(inputs, kernel, group_sizes):
...
+
     # Currently, we only support data and tensor parallelism with Megablox.
     # We all gather the input activations over tensor parallelism to follow strategy
     # in https://parsa.epfl.ch/course-info/cs723/papers/Megatron.pdf.
     input_partition_spec = nn.logical_to_mesh_axes(("activation_batch", None, None))
     gate_logits_pspec = nn.logical_to_mesh_axes(("activation_batch", None, None))
-    w0_pspec = nn.logical_to_mesh_axes((None, None, "mlp"))
-    w1_pspec = nn.logical_to_mesh_axes((None, None, "mlp"))
-    wo_pspec = nn.logical_to_mesh_axes((None, "mlp", None))
+    w0_pspec = nn.logical_to_mesh_axes(("exp", None, "mlp"))
```
Comment (gobbleturk) [inline MaxText/layers/linears.py:520]: Note: For TP_transpose this will have to be modified. The embed dim should be AG over FSDP but not over TP_trasnpose

Doesn't have to be fixed in this PR - I've noted it b/404577816

**Review by gobbleturk — APPROVED** [review]
LGTM! Thanks for the tests!

File: MaxText/layers/linears.py (hunk 1 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:433]: Sure thing!

File: MaxText/layers/linears.py (hunk 2 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:520]: Oh, thanks for the heads-up!

File: MaxText/layers/linears.py [hunk 3]
```diff
@@ -463,14 +508,22 @@ def gmm(inputs, kernel, group_sizes):
         output = output[: hs_shape[0]]
       return output
 
+    def get_all_to_all_params(all_shards_group_sizes, local_expert_size, num_expert_shards):
```
Comment (gagika) [inline MaxText/layers/linears.py:511]: Will you need transform_array function and TransformStrategy for other use cases except this function?

If they are only needed for this function, I suggest doing everything directly here with comments?

**Review by gagika — APPROVED** [review]
Thanks Ran, I have one comment.

File: MaxText/layers/linears.py (hunk 3 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:511]: Sounds good! So far, it only used in this case. Let me wrap those into this function.
